[Step1] Scroll 관련 UI 요소: 지연로딩, 무한스크롤, 횡스크롤박스, 스크롤스파이

ericagong opened the pull request:
## 프로젝트 개요 

안녕하세요. 멘토님 👋  
드디어, 새로운 프로젝트를 시작해 보았습니다 🥳 

- 프로젝트명: UIComponent (가제)
- 프로젝트 핵심 내용: `실무에서 자주 사용되는 UI 요소 구현` 
  - 구현 UI 요소: 
    - 아코디언, 탭 메뉴, 툴팁
    - 반응형 텍스트 박스, 여러줄 말줄임(Line Clamp)
    - 지연 로딩, 무한 스크롤, 횡 스크롤박스, 스크롤 스파이
    - 스낵바, 모달, 팝오버
    - 이미지 슬라이드, 캐러셀, 갤러리
    - 셀렉트 박스, 자동완성, 드래그 앤 드랍 리스트
- 프로젝트 목적: 
  - 동일 UI 컴포넌트를 다양한 방식으로 구현 
  - 각 방식의 장단점, 한계 및 trade-off 분석
  - 주관적으로 현 시점 최적의 구현 방식이 무엇인지 선정 + 의사 결정 과정 [WIKI](https://github.com/ericagong/UIComponent/wiki/UI-Component) 문서화
- 프로젝트 진행 순서: 
  -  step1. React로 각 컴포넌트 구현
    - 컴포넌트 유형별 PR 생성
    - 컴포넌트 별 구현 방식 장단점, 한계 분석, 최적안 선택 -> WIKI 문서화
  - step2. (선택) Vanilla JS로 각 컴포넌트 재구현
  - step3. (필수) 리팩토링 및 배포
    - 재사용성/범용성 고려 -> Headless UI / Compounded Compononent 방향 리팩토링
    - 스토리북 도입 고려
    - MonoRepo 도입 후 타 프로젝트 활용 여부 고려


## PR 요약

이번주는 크게 두 가지 작업을 진행했습니다. 
1. 프로젝트 환경 설정
2. 스크롤 관련 UI 요소 구현 

### 1. 프로젝트 환경 설정
- NextJS + TypeScript 기반 빌드
- SCSS + CSS module 기반 스타일 작성
- 디렉토리 구조: 
  ```bash
  /app
  ├── [...item]/page.tsx          # /[...item] 경로에 해당하는 페이지 뷰 (routes의 key에 매핑된 컴포넌트 렌더링)
  ├── layout.tsx                  # 기본 HTML 구조를 담당하는 레이아웃
  ├── page.tsx                    # 루트(/) 페이지 뷰 (README.md를 렌더링)
  ├── global.scss                 # 앱 전역 스타일
  └── gnb.tsx                     # 좌측 메뉴(GNB: Global Navigation Bar) 컴포넌트
  
  /components
  └── [ComponentName]/            # UI 요소 컴포넌트 (예: Infinite Scroll 등)
      ├── 1_방법A.tsx             # 구현 방법 1번 (예: Scroll Event 기반 등)
      ├── 2_방법B.tsx             # 구현 방법 2번 (예: IntersectionObserver 기반 등)
      └── useLogic.ts             # 로직을 캡슐화한 커스텀 훅 (ex. useScrollSpy)
  └── vanillaWrapper.ts           # 독립적인 VanillaJS 환경을 위한 wrapper 컴포넌트
  
  routes.ts                       # 전체 라우팅 구성 정의 (경로 ↔ 컴포넌트 매핑)

  ```

### 2. 스크롤 관련 UI 요소 구현
지연 로딩, 무한 스크롤, 횡 스크롤박스, 스크롤 스파이 컴포넌트를 구현하였습니다. 
각 UI 요소 구현 방법의 장단점, 한계점, trade-off를 분석하여 최적안을 채택하였습니다. 

- **각 요소의 UX 측면 / 구현 방향성 측면 개선 방향 관련한 피드백**을 부탁드립니다. 
- 이번 PR에서는 **로직과 뷰를 분리**하는데 신경을 써 보았습니다. 
- **관심사에 따라 hook을 분리할 때, 강결합 되거나 재사용성이 낮은 경우 어디까지 분리해야할지에 대한 토의를 나눠보고 싶습니다. (저는 우선 hook 분리가 복잡성을 지나치게 늘리는 경우는 단일 hook으로 처리했습니다.)**


## 각 UI 요소 관련 세부 구현 사항
### ☑️ 지연 로딩 (Lazy Loading)
- 핵심 구현 방법:
  - 뷰포트 내 타겟 이미지 요소의 교차 여부 판단 후 src 속성 주입
  - 교차 이전엔 skeleton UI로 비어있는 이미지 영역 처리
  - 세 가지 방식(getBoundingClientRect, IntersectionObserver, loading 속성) 각각 구현
- 최종 선택안:  loading 속성을 기본 + 미지원 브라우저에 대해 IO 기반 폴리필 적용(하이브리드)
  - loading 속성이 가장 간단하고 직관적 BUT [브라우저 지원 제한적](https://caniuse.com/?search=loading)
  - IntersectionObserver는 메인 쓰레드를 점유 X 백그라운드에서 교차 여부 감지 → 성능 우수

### ☑️ 무한 스크롤
- 핵심 구현 방법:
  - useMockFetchAPI hook으로 fetchAPI 호출 환경 모방
  - 이벤트 핸들러 방식: scroll, resize 이벤트 발생 시, 스크롤 높이와 전체 문서 높이를 직접 계산해 데이터 요청
      - 이벤트 발생 빈도가 많아 throttle 등 최적화 필수
  - Intersection Observer 방식: 최하단 타겟 요소가 뷰포트와 교차 여부 비동기 감시하여 데이터 요청
- 최종 선택안: Intersection Observer 방식 채택
  - Intersection Observer가 메인 쓰레드 부하가 없어 성능 측면 우수 + 구현 편의성 

### ☑️ 횡 스크롤 박스
- 핵심 구현 방법:
  - 좌우 버튼 클릭 시, scrollIntoView로 스크롤 이동 처리
  - IntersectionObserver로 리스트 양 끝 sentinel 요소 교차 여부 감시 ->  좌우 이동 버튼 보임 속성 처리
- 최종 선택안: Intersection Observer 방식 채택

### ☑️ 스크롤 스파이
- 핵심 구현 사항:
  - scrollTop 기반 현재 보이는 콘텐츠 항목을 계산 -> 네비게이션바 active 항목 반영
  - navButton 클릭 -> 해당 콘텐츠 위치로 스크롤 이동 + 네비게이션바 가로 스크롤 자동 가운데 정렬
  - ResizeObserver 추가해 브라우저 크기 변화 감지 기능 추가
- 최종 선택안: Intersection Observer 방식 채택
  - Intersection Observer + Resize Event Handler 중심 구현: 성능 및 구현 편의성 우수
  - Scroll Event + Resize Observer 방식: 발생 빈도 throttle 수동 조절 필요, 메인 쓰레드 병목 발생, 구현 복잡성 높아 비권장
  
 ## 질문
 
  1. lazyLoading/4.polyfill의 폴리필 적용 방식이 적절한지, 폴리필에서 어느 기준까지(실험적인 기능만?) fallback 처리를 해두는게 좋은지?
  
  2. 스크롤 이벤트 핸들러 기반 구현 시, throttle을 적용해 이벤트 핸들러 호출 빈도를 조절하였습니다. 
  이 때 throttled된 함수가 lazyLoading에서는 기대한 방식대로 동작하나, infiniteScroll에서는 최초 1회 이후 호출이 되지 않는 이슈가 발생했습니다. 
  `stale closure` 때문인지? 원인을 파악하고 싶습니다. 
  추가적으로 리액트 내에서 함수나 변수의 동기화를 할 수 있는 방법(`useEvent`, `useSyncExternalStore`) 등에 관해 논의해보고 싶습니다. 
  
  3. scrollSpy/2.IO에서 `Resize Observer` 적용하는 것과 `Resize` 이벤트 핸들러를 등록하는 것 중 더 나은 선택이 무엇인지 토의해보고 싶습니다.

ericagong committed:
feat: 이미지 지연 로딩(Lazy Loading) 구현

핵심 구현 방법:
- 뷰포트 내 타겟 이미지 요소의 교차 여부 판단 후 src 속성 주입
- 교차 이전엔 skeleton UI로 비어있는 이미지 영역 처리
- 세 가지 방식(getBoundingClientRect, IntersectionObserver, loading 속성) 각각 구현

최종 선택안:
- loading 속성이 가장 간단하고 직관적이지만 브라우저 지원이 제한적 (https://caniuse.com/?search=loading)
- IntersectionObserver는 메인 쓰레드를 점유하지 않고 백그라운드에서 교차 여부 감지 → 성능 우수
- loading 속성을 기본으로, 미지원 브라우저에 대해 IO 기반 폴리필 적용하는 하이브리드 전략 채택

ericagong committed:
feat: 무한 스크롤(Infinite Scroll) 구현

핵심 구현 방법:
- useMockFetchAPI hook으로 fetchAPI 호출 환경 모방
- 이벤트 핸들러 방식: scroll, resize 이벤트 발생 시, 스크롤 높이와 전체 문서 높이를 직접 계산해 데이터 요청
    - 이벤트 발생 빈도가 많아 throttle 등 최적화 필수
- Intersection Observer 방식: 최하단 타겟 요소가 뷰포트와 교차 여부 비동기 감시하여 데이터 요청

최종 선택안:
- Intersection Observer가 메인 쓰레드 부하가 없어 성능 측면 우수
- Intersection Observer가 구현 편의성 우수

ericagong committed:
feat: 수평 스크롤 박스(Horizontal ScrollBox) 구현

핵심 구현 방법:
- 좌우 버튼 클릭 시, scrollIntoView로 스크롤 이동 처리
- IntersectionObserver로 리스트 양 끝 sentinel 요소 교차 여부 감시 ->  좌우 이동 버튼 보임 속성 처리

ericagong committed:
refactor: 로직과 뷰 분리

    핵심 변경 사항
    - useIntersectionObserver: 단일 entry -> entries 배열로 로직 확장
    - horizontalScrollBox: useButtonEnabled, useScrollToEdge로 관심사에 따라 로직 분리
    - lazyLoading: useLoading, useLazyLoadHandler, useLazyLoad로 관심사에 따라 로직 분리

ericagong committed:
feat: Scroll Spy 구현

  핵심 구현 사항:
  - scrollTop 기반 현재 보이는 콘텐츠 항목을 계산 -> 네비게이션바 active 항목 반영
  - navButton 클릭 -> 해당 콘텐츠 위치로 스크롤 이동 + 네비게이션바 가로 스크롤 자동 가운데 정렬
  - ResizeObserver 추가해 브라우저 크기 변화 감지 기능 추가

  최종 선택안:
  - Intersection Observer + Resize Observer 중심 구현: 성능 및 구현 편의성 우수
  - Scroll Event 기반 계산 방식: 발생 빈도 throttle 수동 조절 필요, 메인 쓰레드 병목 발생, 구현 복잡성 높아 비권장

JunilHwang commented on src/components/hook/useIntersectionObserver.ts:
```diff
@@ -0,0 +1,55 @@
+
+        return visibleEntries;
+      });
+    };
```
흠.. 개인적으로 훅 내부에서 상태를 가지기보단, 바깥에 이벤트르 통해 알려주는 방식으로 만들면 어떨까 싶어요.

```tsx
useIntersectionObserver(targetsRef, { onIntersect: (entires) => console.log(entries)  })
```

요로코롬?

이렇게 했을 때 intersection을 통해 할 수 있는 것들이 더 많아지지 않을까!? 라는 생각이 듭니다 ㅎㅎ

JunilHwang commented on src/components/lazyLoading/3_loadingAttribiute/index.tsx:
```diff
@@ -0,0 +1,47 @@
+import cx from '../cx';
+import data from '../data';
+
+export const LazyImage = ({
+  src,
+  width,
+  height,
+}: {
+  src: string;
+  width: number;
+  height: number;
+}) => {
+  const imgRef = useRef<HTMLImageElement>(null);
+  const { loading, setLoaded } = useLoading(true);
+
+  return (
+    <img
+      ref={imgRef}
+      className={cx({ loading: loading })}
+      alt=''
+      width={width}
+      height={height}
+      src={src}
+      loading='lazy'
+      onLoad={setLoaded}
+    />
+  );
+};
```
lazyImage가 여기저기서 쓰여서 공통 컴포넌트로 만들어서 관리해주면 좋지 않을까!? 라는 생각이 들어요!

ericagong replied:
### ☑️  질문드립니다 

혹시 1, 2 중 어떤 맥락의 피드백일까요? 
1. lazyLoading 하위 폴더들(1_eventHandler, 2_intersectionObserver, 3_loadingAttribute)에서 사용하는 lazyImage를 공통 컴포넌트화해라
2. 외부에서도 lazyImage를 사용하니 별도의 파일로 분리해라

만약 1안이라면,
@componenet/lazyLoading 내부의 1_eventHandler, 2_intersectionObserver, 3_loadingAttribute 내부의 `lazyImage` 컴포넌트는 지금 구현 방식에 따라서 내부적으로 다 다른 로직을 가지고 있어 별도의 컴포넌트로 선언했습니다. 

2안이라면, 이미 `export const` 형식으로 필요한 경우 내보내기하여 사용하고 있는데, 별도의 파일로 분리해 `export default`하는 경우가 더 명확하기 때문에 제안해주신 걸까요?

JunilHwang replied:
1안에 대한 이야기였어요!

> @componenet/lazyLoading 내부의 1_eventHandler, 2_intersectionObserver, 3_loadingAttribute 내부의 lazyImage 컴포넌트는 지금 구현 방식에 따라서 내부적으로 다 다른 로직을 가지고 있어 별도의 컴포넌트로 선언했습니다.

어떤 내용인지 이해했습니다 ㅎㅎ 
이름만 똑같지 다 다른 기능이군요..!

JunilHwang replied:
멘토링 시간에 나온 이야기: 네임스페이스를 활용해도 좋을 것 같음!

https://www.typescriptlang.org/ko/docs/handbook/namespaces.html#%EB%8B%A4%EC%A4%91-%ED%8C%8C%EC%9D%BC-%EB%84%A4%EC%9E%84%EC%8A%A4%ED%8E%98%EC%9D%B4%EC%8A%A4-multi-file-namespaces

JunilHwang commented on src/components/lazyLoading/3_loadingAttribiute/index.tsx:
```diff
@@ -0,0 +1,47 @@
+  width: number;
+  height: number;
+}) => {
+  const imgRef = useRef<HTMLImageElement>(null);
```
imgRef를 만들긴하는데, 쓰이질 않고 있네요!?

ericagong replied:
변경해두었습니다. 
다음 PR 부터는 점검할 수 있도록 프로젝트에 eslint 설정을 해두겠습니다.

JunilHwang commented on src/components/lazyLoading/3_loadingAttribiute/index.tsx:
```diff
@@ -0,0 +1,47 @@
+import cx from '../cx';
+import data from '../data';
+
+export const LazyImage = ({
+  src,
+  width,
+  height,
+}: {
+  src: string;
+  width: number;
+  height: number;
+}) => {
+  const imgRef = useRef<HTMLImageElement>(null);
+  const { loading, setLoaded } = useLoading(true);
```
바깥에서 onLoad 함수도 같이 받아오면 어떨까 싶어요!
더 정확히는 img의 모든 props를 받아올 수 있는 구조면 어떨까요!?

ericagong replied:
LazyImage를 더 범용적인 컴포넌트로 만들기 위해, img의 모든 props를 수용하고 onLoad도 외부 주입 가능하도록 확장하였습니다.

☑️ AS-IS
- LazyImage는 src, width, height만 받음
- onLoad은 useLoading 훅에서 직접 내부 정의된 setLoaded만 사용
- 따라서 바깥에서 onLoad 같은 추가 핸들링을 하고 싶어도 못함

☑️ TO-BE 
- LazyImage가 img의 모든 props (alt, style, onClick, onLoad, ...)를 받을 수 있도록 확장
- onLoad도 외부에서 전달된 콜백과 내부의 setLoaded를 함께 실행할 수 있도록 조합


✅ Question
- "유연한 컴포넌트 설계란 어디까지 열고 어디까지 닫아야 하는가?"에 대해 고민해 보았습니다. 
- `OCP` 원칙 = 확장에는 열려있고, 변경에는 닫혀있어야 한다. 를 생각해 봤을 때,  
- **핵심 로직은 닫고, 입력(props)은 열어두는 것**이라고 판단했습니다. 
- 그래서 핸들러 전체를 열어두되, 컴포넌트 내부의 핵심 로직을 먼저 실행하고, 외부에서 전달된 핸들러는 보조적으로 실행하도록 코드를 변경하였습니다. 
```javascript
 const handleLoad: React.ReactEventHandler<HTMLImageElement> = (e) => {
  setLoaded(); // 내부가 우선
  onLoad?.(e); // 외부는 선택적으로
};
 ```
 - 여기서, **"그럼 어떤 핸들러까지 열어주는게 맞을까"**라는 고민이 들어 질문드립니다. 
   - 저는 img 태그에서 지금 `onLoad`, `onError`, `onClick` 핸들러를 열어두었습니다. 
   - 외부 핸들러를 언제 열어둘 것인가?에 대한 제 나름대로의 판단 기준을 정리해 보았습니다. 
   - **멘토님의 판단 기준, 실무에서의 판단 관행이 있다면 이야기 나눠보고 싶습니다.** 
   
## 🤔 외부 핸들러 열기 판단 기준 3가지
**핵심 아이디어: 외부에 열어줄 핸들러는 확장 가능한 비즈니스 로직 영역만!
내부의 중요한 핵심 동작은 절대 외부에 위임하지 않고 내부에서 선제 처리!**

1. 내부 로직이 끝난 후 외부 확장이 필요할 수 있는가?
- ex: 로딩 완료 후 애니메이션, 로깅, 상태 변화
2. 핸들러가 UI/UX에 따라 다양하게 달라질 수 있는가?
- ex: onClick, onError 같은 건 비즈니스 로직마다 다름
3. 핸들러의 동작이 내부 기능과 충돌하지 않는가?
- 내부 기능과 충돌하거나 우선순위가 꼬일 수 있는 건 제한

JunilHwang replied:
범용적인 컴포넌트일수록 최대한 많이 열어줘야 한다고 생각해요!
특히 이미지 같은 경우에는 어떤 상황에서든 사용될 수 있는 그런 컴포넌트라서요 ㅎㅎ

모든 컴포넌트에 동일한 규칙을 정의하기보단, 컴포넌트가 쓰이는 맥락을 고려해야할 것 같아요!

JunilHwang commented on src/components/lazyLoading/4_polyfill/index.tsx:
```diff
@@ -0,0 +1,43 @@
+  } else {
+    return <LazyImageIO src={src} width={width} height={height} />;
+  }
+};
```
오호.. 이런 폴리필도 만들어주셨군요!

ericagong replied:
### 💥 문제 상황: 이 코드가 SSR에서 평가되면?
```typescript
const supportLazyLoading =
  typeof window !== 'undefined' && 'loading' in HTMLImageElement.prototype;
```
- 문제 발생 원인: 코드 실행 시점 = 컴포넌트 렌더링 시점
  - 따라서 그 시점이 서버인지 클라이언트인지에 따라 코드가 다르게 평가됨
  - SSR에서 평가되면 
    - `typeof window !== 'undefined' → ❌ false` (서버에는 window 없음)
    - 항상 LazyImageIO로 떨어짐 (IntersectionObserver fallback만 사용됨)
    - 즉, 브라우저가 native lazy loading을 지원해도, 클라에선 못쓰는 상태가 됨

### ✅ 해결 방법: 클라이언트 전용 렌더링 강제
```typescript
export const LazyImageWrapper = (props: LazyImageProps) => {
  const [supportLazy, setSupportLazy] = useState<boolean | null>(null);

  useEffect(() => {
    const supported = 'loading' in HTMLImageElement.prototype;
    setSupportLazy(supported);
  }, []);

  if (supportLazy === null) {
    return <div className="spinner" />; // 초기 상태 로딩 스피너
  }

  return supportLazy ? (
    <LazyImageNative {...props} />
  ) : (
    <LazyImageIO {...props} />
  );
};
```
- 핵심 아이디어: `useEffect + state`로 조건 분기
  - useEffect는 클라이언트에서만 실행되므로 안전
  - 단, 첫 렌더링 시 아무것도 안 나올 수 있으니, 그 때 로딩 스피너 넣어두기
  - Next.js의 next/dynamic으로 컴포넌트를 CSR 전용으로 처리할 수도 있지만,
useEffect + state 방식은 렌더링 시점을 명확히 예측할 수 있고,
UI와 로직을 같은 컴포넌트 안에서 일관되게 관리할 수 있다는 점에서 더 직관적이므로 이 방식을 채택

JunilHwang commented on src/components/lazyLoading/4_polyfill/index.tsx:
```diff
@@ -0,0 +1,43 @@
+  threshold: 0,
+};
+
+export const LazyImage = ({
+  src,
+  width,
+  height,
+}: {
+  src: string;
+  width: number;
+  height: number;
+}) => {
+  const supportLazyLoading =
+    typeof window !== 'undefined' && 'loading' in HTMLImageElement.prototype;
+
+  if (supportLazyLoading) {
+    return <LazyImageNative src={src} width={width} height={height} />;
+  } else {
+    return <LazyImageIO src={src} width={width} height={height} />;
```
이게 SSR 환경에서 하이드레이션을 할 때 문제될 수 있을 것 같아요..!

JunilHwang replied:
서버에서는 LazyImageIO로 항상 렌더링 되고, 클라에서는 뭘로 렌더링 될지 모르는..

ericagong replied:
앞선 코멘트에서 수정해두었습니다!

JunilHwang commented on src/components/hook/useStable.ts:
```diff
@@ -0,0 +1,23 @@
+  // 리렌더마다 최신 함수 갱신
+  useEffect(() => {
+    fnRef.current = fn;
+  }, [fn]);
```
저도 이거 똑같이 한 번 만들어봤는데,
useEffect 시점에 최신화를 하게 되면 최초에 렌더링 되는 시점에 fn의 값이 달라지는..? 그런 문제가 있어요.

그래서 


```suggestion
  const fnRef = useRef<T>(fn);
  fnRef.current = fn;
```

이렇게 렌더링할 때 마다 최신화를 해줘야 하네요

JunilHwang commented on src/components/hook/useThrottle.ts:
```diff
@@ -0,0 +1,17 @@
+  }, [delay]);
+
+  return throttledFn;
+}
```
겸사겸사 debounce도 만들어주세요!ㅋㅋㅋ

JunilHwang commented on src/components/hook/useThrottle.ts:
```diff
@@ -0,0 +1,17 @@
+import { useRef, useEffect, useCallback } from 'react';
+import useStable from '@/components/hook/useStable';
+
+export default function useThrottle(fn: () => void, delay: number = 1000) {
```
```suggestion
export default function useThrottle(fn: () => void, delay = 1000) {
```
기본값을 통해 타입이 추론되서, 이럴 때는 타입을 정의해주지 않아도 무방하답니다!

JunilHwang commented on src/components/hook/useThrottle.ts:
```diff
@@ -0,0 +1,17 @@
+      lastCallRef.current = now;
+      stableFn();
+    }
+  }, [delay]);
```
아싸리 이 함수도 useStable로 감싸줘도 좋을 것 같아요!
사실 모든 useCallback을 다 대체해도 좋지 않을까 싶은...

JunilHwang commented on src/components/horizontalScrollBox/hooks/useScrollToEdge.ts:
```diff
@@ -0,0 +1,50 @@
+    };
+  };
+
+  const scrollToEdge = (direction: ButtonDirection) => {
+    const { $leftItem, $rightItem } = getVisibleItemsOnEdge();
+
+    const $target = direction === 'prev' ? $leftItem : $rightItem;
+
+    $target?.scrollIntoView({
+      behavior: 'smooth',
+      block: 'nearest',
+      inline: direction === 'prev' ? 'end' : 'start',
+    });
+  };
+
+  return scrollToEdge;
```
이 함수는 순수함수로 분리해줘도 좋을 것 같네요 ㅎㅎ

ericagong replied:
분리해두었습니다.

###  🤔 순수함수로 분리해야할 순간을 어떻게 캐치할까?
- 순수함수 = 입력만으로 결과가 결정되고, 사이드 이펙트 없이 동작하는 함수
- 재사용성 향상, 테스트 용이성, 의존성 분리, 유지 보수성 향상... -> 순수함수를 지향해야 하는 이유는 수없이 많다고 생각합니다. 
- 하지만 막상 코드를 작성하다보면, 기능 구현에 치중하다보니 순수함수 분리를 놓치는 것 같아요.
- 그래서 저만의 순수함수 분리 기준을 세워보았습니다. 멘토님의 기준이 있다면 이야기 나눠보고 싶습니다. 

### ✅  순수함수 분리 시그널
- 입출력이 명확한가?
- 외부 상태 없이 작동하는가?
- 다른 곳에서 반복되거나 확장될 가능성이 있는가?

JunilHwang commented on src/components/horizontalScrollBox/hooks/useScrollToEdge.ts:
```diff
@@ -0,0 +1,50 @@
+      $leftItem: $items[leftIndex],
+      $rightItem: $items[rightIndex],
+    };
+  };
```
심심하면(?) 이런 코드에 대한 테스트도 추가해보시면 좋겠어요!
막 너무 타이트하게 라기보단, 간단하게 검증하는 용도랄까..

JunilHwang commented on src/components/infiniteScroll/1_EventHandler/useInfiniteScroll.ts:
```diff
@@ -0,0 +1,45 @@
+const useInfiniteScroll = () => {
+  const { status, data, fetchMore } = useMockFetchAPI();
+
+  const handleInfiniteScroll = useCallback(() => {
+    // console.count('handleInfiniteScroll called');
+    if (status === 'LOADING') return;
+
+    const scrollTop = window.scrollY || document.documentElement.scrollTop;
+    const viewportHeight = window.innerHeight;
+    const documentHeight = document.documentElement.scrollHeight;
+
+    if (documentHeight - (scrollTop + viewportHeight) <= MORE_THRESHOLD) {
+      fetchMore();
+    }
+  }, [status, fetchMore]);
```
useStable을 이 함수에 씌워주면 좋겠어요!

JunilHwang commented on src/components/infiniteScroll/1_EventHandler/useInfiniteScroll.ts:
```diff
@@ -0,0 +1,45 @@
+  //   DELAY,
+  // );
+
+  useEffect(() => {
+    window.addEventListener('scroll', handleInfiniteScroll);
+    window.addEventListener('resize', handleInfiniteScroll);
+
+    // 최초 렌더링 시 트리거
+    handleInfiniteScroll();
+
+    return () => {
+      window.removeEventListener('scroll', handleInfiniteScroll);
+      window.removeEventListener('resize', handleInfiniteScroll);
+    };
+  }, [handleInfiniteScroll]);
```
이 함수가 useStable로 만들어놓은 함수를 의존하는 경우, 최초에 한 번만 useEffect에 함수를 등록하면 되서 좋답니다 ㅎㅎ
지금은 status와 fetchMore이 변경될 대 마다 이벤트를 제거하고 다시 등록하는 방식이네요.

JunilHwang commented on src/components/infiniteScroll/utils.ts:
```diff
@@ -0,0 +1,41 @@
+const randomShuffle = <T>(arr: T[]): T[] => {
+  // Fisher-Yates shuffle algorithm
+  const result = [...arr];
+  for (let i = result.length - 1; i > 0; i--) {
+    const j = Math.floor(Math.random() * (i + 1));
+    [result[i], result[j]] = [result[j], result[i]];
+  }
+  return result;
```
```suggestion
  return [...arr].sort(() => Math.random() - 0.5);
```
요로코롬 할 수도 있어요!

ericagong replied:
아! 여기서 저는 `.sort(() => Math.random() - 0.5)`가 균등 확률을 보장할 수 없어서...
정형화된 확률적 셔플 알고리즘인 `Fisher-Yates` 방식의 코드를 채택했습니다.

### 💣 .sort(() => Math.random() - 0.5)의 문제
- 자바스크립트의 Array.prototype.sort()는 브라우저마다 내부적으로 다른 정렬 알고리즘을 사용함
  - V8 (Chrome, Node.js): TimSort 기반
  - SpiderMonkey (Firefox): MergeSort 기반
  - 이 알고리즘들은 정렬 안정성, 성능 최적화가 목적이지, 무작위성을 보장하지 않음
- sort()는 내부적으로 비교-교환하는 과정이 있고, 순서를 바꾸는 패턴이 정해져 있음
  - 따라서 이 과정에 랜덤한 비교값을 넣는다고 해도,
    - 일부 순열이 더 자주 나오는 통계적 편향이 생기고, 모든 순열이 동일 확률로 나올 보장이 없음
- `() => Math.random() - 0.5` 또한 정렬 기준으로는 적합한 함수가 아님
  - 값의 일관성과 방향성이 없기 때문에 정렬 결과가 불확정적이며, 편향된 순열이 생기기 때문

### 🚀 Fisher-Yates 셔플을 사용한 이유
- Fisher-Yates = 정형화된 확률적 셔플 알고리즘
  - 확률 균등성 보장 = 각 순열이 1/n! 확률로 등장
  - 독립성 보장 = 정렬 알고리즘에 종속되지 않음
  - 일관성 보장 = 어디서 실행해도 동일한 셔플 품질 보장
- Fisher-Yates의 구조를 모방하여 코드 작성
  - 뒤에서부터 순차적으로 내려오며
  - 현재 인덱스(i)보다 작거나 같은 범위에서 랜덤하게 하나(j)를 뽑아
  - arr[i]와 arr[j]를 스왑

JunilHwang commented on src/components/lazyLoading/1_EventHandler/useLazyLoadHandler.tsx:
```diff
@@ -0,0 +1,23 @@
+  return rect.top < window.innerHeight && rect.bottom > 0;
+};
+
+const useLazyLoadHandler = (targetRef: RefObject<HTMLImageElement>) => {
+  const handleLazyLoad = () => {
+    const $img = targetRef.current;
+    // 이미 로드된 target 미처리
+    if (!$img || $img.getAttribute('src')) return;
+
+    if (isIntersecting($img)) {
+      $img.setAttribute('src', $img.dataset.src || '');
+    }
+  };
+
+  return { handleLazyLoad };
+};
```
흠 이름이 lazyLoad인데, 실제로 하는일은 "뷰포트에 진입하면 로딩해줘!" 라서

"loadIntersected"처럼 표현해주면 어떨까 싶어요!

```tsx
const useLoadIntersected = (targetRef: RefObject<HTMLImageElement>) => {
  return () => {
    const $img = targetRef.current;
    // 이미 로드된 target 미처리
    if (!$img || $img.getAttribute("src")) return;

    if (isIntersecting($img)) {
      $img.setAttribute("src", $img.dataset.src || "");
    }
  };
};
```

요로코롬...

ericagong replied:
lazyLoad는 개념과 관련된 명칭이라 추상적이었다면,
loadIntersected는 함수가 하는 행위의 본질을 더 잘 설명하는 것 같아요!

사실 함수 네이밍은 늘 어려운 문제라고 생각하는데요.
저는 나름대로 함수 네이밍 3원칙을 가지고 이름을 지으려고 하는 편입니다...(갈 길이 멀지만...)
**혹시 멘토님의 함수 이름 짓기 요령이 있다면 전수해주세요...**

### ✅ 함수 네이밍 3원칙
1. 행위 중심 = “무엇을 하는가”를 말하라
- 함수 이름은 반드시 **행위(동사)**로 시작하고, 무엇을 하는지 드러나야 한다.
- ✅ 좋은 예: loadImage, scrollToTop, toggleTheme
- ❌ 나쁜 예: intersectionObserverLoader
→ 이건 어떤 기술로 구현됐는지는 보여주지만, 무슨 일을 하는지는 안 보임.

2. 의도는 구체적으로, 구현은 감춰라
- 이름엔 목적만 드러내고, 내부의 **구현 방식(Observer, fetch 등)**은 숨겨라.
- 구현은 언제든 바뀔 수 있지만, 의도는 그대로 유지되어야 하기 때문!
- ✅ 좋은 예: useIsInViewport, useSaveDraft
- ❌ 나쁜 예: useResizeObserverForHeight
→ 실제로 height를 detect하는 방식은 resize observer가 아닐 수도 있다. 이름에 기술이 묻어나면 나중에 바꾸기 어려워짐.

3. 이름만 봐도 사용처가 떠오르게
- 함수 이름은 호출되는 맥락을 함께 상상할 수 있어야 한다.
- 맥락이 드러나는 구체적인 표현을 써야 헷갈리지 않는다.
- ✅ 좋은 예: useAutoFocus, getFormErrors, submitLoginForm
- ❌ 나쁜 예: doStuff, handleThing, processData
→ 이런 이름은 “뭔가 하긴 하는데” 어디에 쓰는지 추측이 안 됨.

### ✅ 함수 네이밍 체크 리스트
- 이 함수는 “무엇”을 하는가?
- 이 이름은 구현 방식을 드러내고 있진 않은가?
- 이 이름만 보고, 다른 개발자들도 어디에 쓰일지 감을 바로 잡을 수 있는가?

JunilHwang commented on src/components/lazyLoading/2_IntersectionObserver/useLazyLoad.tsx:
```diff
@@ -0,0 +1,25 @@
+
+const IOOptions: IntersectionObserverInit = { threshold: 0 };
+
+const useLazyLoad = (imgRef: RefObject<HTMLImageElement>, src: string) => {
+  const { observerRef, visibleEntries } = useIntersectionObserver(
+    imgRef,
+    IOOptions,
+  );
+
+  useEffect(() => {
+    const entry = visibleEntries[0];
+    if (entry?.isIntersecting) {
+      const $img = imgRef.current;
+      if ($img) {
+        $img.setAttribute('src', src);
+        observerRef.current?.disconnect();
+      }
+    }
+  }, [src, visibleEntries]);
+};
+
+export default useLazyLoad;
```
앞선 리뷰와 동일합니다!

JunilHwang commented on src/components/scrollSpy/1_EventHandler/useScrollSpy.ts:
```diff
@@ -0,0 +1,111 @@
+      left: scrollLeft,
+      behavior: 'smooth',
+    });
+  });
```
useEffect에 dependencies 명시가 안 되어있네요..!

JunilHwang commented on src/components/scrollSpy/1_EventHandler/useScrollSpy.ts:
```diff
@@ -0,0 +1,111 @@
+    const scrollTop = document.scrollingElement?.scrollTop || 0;
+    contentItemsRef.current?.forEach(($item, index) => {
+      if (!$item) return;
+      const { top, height } = $item.getBoundingClientRect();
```
top, height 는 최초에 한 번만 계산해도 좋지 않을까 싶네요!?
scrollTop은 렌더링될 때 마다 필요하지만, 현재 렌더링된 요소들의 top, height는 이미 고정되어있을 것 같아서요!

JunilHwang commented on src/components/scrollSpy/2_IntersectionObserver/useScrollSpy.ts:
```diff
@@ -0,0 +1,87 @@
+  threshold: 0.5,
+};
+
+const useScrollSpy = () => {
```
1_EventHandler 의 훅과 많이 겹치는 느낌인데,
재활용할 수 있는 부분은 순수 함수로 분리해줘도 좋을 것 같아요!

JunilHwang reviewed (commented):
ㅎㅎ 다양한 훅을 구현해주셔서 학습이 꽤 많이 되었을 것 같아요!
이런 프로젝트도 좋네요!

ericagong committed:
feat: Accordion 구현

 핵심 구현 사항:
  - SEO 고려 -> 콘텐츠 항상 DOM 트리 포함 처리
  - 접근성 고려 -> WAI-ARIA 속성 부여
  - ctrl+F 검색 UX 지원 -> hidden=until-found 속성 사용 OR details/summay 태그 기반 구현

 최종 선택안: 단일 열림 정책 아코디언 -> 커스텀 구현 && 다중 열림 정책 아코디언 -> details + summay 기반 구현
  - 단일 열림 정책 아코디언: Accodion Item: Button(.trigger) + Div(.content) 기반 구현
    - 장점:
     - hidden=until-found + beforeMatch event handler 등록 -> ctrl+F UX 구현 가능
     - transition max-height 기반 부드러운 애니메이션 처리 가능
    - 한계:
     - 일부 브라우저 beforeMatch + until-found 미지원
   - 다중 열림 정책 아코디언: details + summary 태그 기반 구현
    - 장점:
     - ctrl+F 자동 처리
     - 구현 용이성
    - 한계:
     - 스타일 재설정 필요 및 단일 열림 정책 아코디언 구현 복잡성

ericagong committed:
feat: tabMenu 구현

  핵심 구현 사항
  - 접근성 준수: WAI-ARIA 속성 작성
  - 키보드 UX 향상: 키보드 방향키 포커스 이동 및 tabInde 통해 키보드 탐색 현재탭 한정
  - 시맨틱 준수

  최종 선택안: display Attribute block <-> none 기반 탭메뉴 구현
  - 특정 panel만 렌더링 시 SEO 최적화 실패
  - 한계: display 속성 사용시, transition 불가 -> 애니메이션 필요시 opacity, visibility, max-height 방식 고려

ericagong committed:
feat: 반응형 텍스트박스 구현

  핵심 구현 방법
  - controlled: 2D Canvas Element 기반 줄 수 계산 -> textarea rows 반영
  - uncontrolled: clone textarea Element 기반 줄 수 측정 -> textarea rows 반영
  - 접근성: screen reader only 클래스로 보조기기 접근성 향상

  최종 선택안: uncontrolled clone Element 기반 반응형 텍스트박스
  - 장점: 오차 없이 정확한 줄 수 계산 가능(CSS와 동일 렌더링 기준: 폰트, 줄바꿈, 줄간격 규칙)
  - 한계: uncontrolled 방식이므로 개발자가 리액트 생태계에 상태 pull 하는 타이밍 수동 조절 + 개발자가 스타일 동기화 필
  -  한계: 모바일 환경에서 키보드 등장 시 줄 수 측정 이슈 발생 가능

ericagong committed:
chore: 폴더 이름 통일

ericagong committed:
feat: LineClamp 구현

  핵심 구현 방법
  - canvas 기반 줄 개수 측정하는 controlled 방식
  - clone Element 기반 줄 개수 측정하는 uncontrolled 방식
  - 접근성 향상: WAI-ARIA 속성 부여

  최종 선택안: clone Element 기반 uncontrolled 방식
  - 장점: 높은 계산 정확도
  - 단점: 성능 저하 가능성 + 리사이즈 대응 별도로 필요
  - 주의사항: 개발자가 clone 요소와 대상 요소 CSS 수동으로 일치시켜야 함

ericagong committed:
feat: tooltip 구현

ericagong commented:
## PR 요약

이번 주에는 아코디언, 탭메뉴, 툴팁, 반응형 Textbox, LineClamp를 구현해 보았습니다. 
각 UI 요소 구현 방법의 장단점, 한계점, trade-off를 분석하여 최적안을 채택하였습니다. 

- **각 요소의 UX 측면 / 구현 방향성 측면 개선 방향 관련한 피드백**을 부탁드립니다. 
- 이번 PR에서도 **로직과 뷰를 분리**하는데 신경을 써 보았습니다. 
- **관심사에 따라 hook을 분리할 때, 강결합 되거나 재사용성이 낮은 경우 어디까지 분리해야할지에 대한 토의를 나눠보고 싶습니다. (저는 우선 hook 분리가 복잡성을 지나치게 늘리는 경우는 단일 hook으로 처리했습니다.)**


## 각 UI 요소 관련 세부 구현 사항
### ☑️ 아코디언(Accordion)
- 핵심 구현 사항:
  - SEO 고려 -> 콘텐츠 항상 DOM 트리 포함 처리
  - 접근성 고려 -> WAI-ARIA 속성 부여
  - ctrl+F 검색 UX 지원 -> hidden=until-found 속성 사용 OR details/summay 태그 기반 구현
- 최종 선택안: 단일 열림 정책 아코디언 -> 커스텀 구현 && 다중 열림 정책 아코디언 -> details + summay 기반 구현
  - 단일 열림 정책 아코디언: Accodion Item: Button(.trigger) + Div(.content) 기반 구현

### ☑️ 탭메뉴(TabMenu)
- 핵심 구현 사항
  - 접근성 준수: WAI-ARIA 속성 작성
  - 키보드 UX 향상: 키보드 방향키 포커스 이동 및 tabInde 통해 키보드 탐색 대상을 현재 탭으로 한정
  - 별도 애니메이션 처리 X
 - 최종 선택안: display Attribute block <-> none 기반 탭메뉴 구현
  - 특정 panel만 렌더링 시 SEO 최적화 실패하므로, display 속성 기반 구현 방식 채택
  - 한계: display 속성 사용시, transition 불가 -> 애니메이션 필요시 opacity, visibility, max-height 방식 고려

### ☑️ 툴팁(Tooltip)
- 핵심 구현 방법:
  - click 기반 툴팁: 페이지 내 단일 툴팁 열림 처리 위해 context API OR details, summary 태그 기반 구현
  - hover 기반 툴팁: flickering 처리 위해 여러 타겟 요소 mouseEnter, mouseLeave 이벤트 핸들러 등록 처리
  - 항상 뷰포트 내에 툴팁 렌더링 보장 -> useAutoPlacement Hook으로 style 계산해서 JS로 입히기
- 최종 선택안: 요구사항에 따라 click 기반 tooltip OR hover 기반 tooltip 구현
  - PC + 간단한 설명 → Hover 기반 구현
  - 모바일 지원 또는 툴팁 내부 상호작용 → Click 기반 + SingleActiveItemContext 등으로 단일 열림 제어

### ☑️ 반응형 텍스트 박스(TextBox)
- 핵심 구현 방법
  - 사용자 기입 텍스트의 줄 수 계산 -> [min, max] 사이에서 반응형으로 textarea row 증감 처리
  - controlled: 2D Canvas Element 기반 줄 수 계산 -> textarea rows 반영
  - uncontrolled: clone textarea Element 기반 줄 수 측정 -> textarea rows 반영
- 최종 선택안: uncontrolled clone Element 기반 반응형 텍스트박스
  - 장점: 오차 없이 정확한 줄 수 계산 가능(CSS와 동일 렌더링 기준: 폰트, 줄바꿈, 줄간격 규칙)
  - 한계: uncontrolled 방식이므로 개발자가 리액트 생태계에 상태 pull 하는 타이밍 수동 조절 + 개발자가 스타일 동기화 필수
             모바일 환경에서 키보드 등장 시 줄 수 측정 이슈 발생 가능

### ☑️ LineClamp
- 핵심 구현 방법
  - canvas 기반 줄 개수 측정하는 controlled 방식
  - clone Element 기반 줄 개수 측정하는 uncontrolled 방식
  - 접근성 향상: WAI-ARIA 속성 부여
 - 최종 선택안: clone Element 기반 uncontrolled 방식
   - 장점: 높은 계산 정확도 보장
   - 단점: 성능 저하 가능성 + 리사이즈 대응 별도로 필요
   - 주의사항: 개발자가 clone 요소와 대상 요소 CSS 수동으로 일치시켜야 함

JunilHwang commented on src/components/accordion/1_hiddenAttribute/index.tsx:
```diff
@@ -0,0 +1,86 @@
+    if (!$contentEl) return;
+
+    if ($contentEl) {
+      $contentEl.addEventListener('beforematch', toggle);
+    }
+    return () => {
+      if ($contentEl) $contentEl.removeEventListener('beforematch', toggle);
+    };
```
오호... 이런 이벤트도 있군요 ㅎㅎ 덕분에 하나 알아갑니다..!

JunilHwang commented on src/components/accordion/1_hiddenAttribute/index.tsx:
```diff
@@ -0,0 +1,86 @@
+
+  useEffect(() => {
+    const $contentEl = contentRef.current;
+    if (!$contentEl) return;
+
+    if ($contentEl) {
+      $contentEl.addEventListener('beforematch', toggle);
+    }
+    return () => {
+      if ($contentEl) $contentEl.removeEventListener('beforematch', toggle);
+    };
+  }, [toggle]);
```
오호... 이런 이벤트도 있군요 ㅎㅎ 덕분에 하나 알아갑니다!

그리고

```suggestion
    if (!$contentEl) return;

    $contentEl.addEventListener('beforematch', toggle);
    return () => $contentEl.removeEventListener('beforematch', toggle);
  }, [toggle]);
```
$container가 없는 상황알 앞에서 필터링해주기 때문에 이렇게 처리해도 무방해보여요!

JunilHwang commented on src/components/accordion/1_hiddenAttribute/index.tsx:
```diff
@@ -0,0 +1,86 @@
+        aria-labelledby={triggerID}
+        aria-hidden={!current}
+        // @ts-ignore
+        HIDDEN={current ? undefined : 'until-found'}
```
근데 useEffect에서 이벤트를 등록해줄 필요가 있을까요? 컴포넌트에 직접 등록해도 되지 않을까 싶어서요!

JunilHwang replied:
아.. 리액트에서 지원을 안 해주는구나 ㅋㅋ;

JunilHwang commented on src/components/accordion/2_detailsSummary/MultiOpenAccordion.tsx:
```diff
@@ -0,0 +1,55 @@
+    return () => {
+      $contentEl.removeEventListener('beforematch', handleOpen);
+    };
+  });
```
```tsx
const currentRef = useBeforeMatch(handleOpen)
```
이렇게 처리할 수 있도록 훅을 하나 만들어줘도 무방해보여요!

JunilHwang commented on src/components/accordion/2_detailsSummary/SingleOpenAccordion.tsx:
```diff
@@ -0,0 +1,67 @@
+    return () => {
+      $panelEl.removeEventListener('beforematch', toggle);
+    };
+  });
```
useEffect에 deps 를 명시해주지 않으면 렌더링이 될 때 마다 이펙트 내부의 함수가 발생하는걸로 알고 있어요..!

JunilHwang commented on src/components/hook/useHover.ts:
```diff
@@ -0,0 +1,50 @@
+  delay = 100,
+) => {
+  const [isHovered, setIsHovered] = useState(false);
+  const isHovering = useRef(false);
```
두 값이 항상 동기화 되고 있는데, 
ref와 state로 분리한 이유가 있을까요?

JunilHwang commented on src/components/tabMenu/1.displayAttribute/index.tsx:
```diff
@@ -0,0 +1,94 @@
+    title: string;
+    current: boolean;
+    handleClick: () => void;
+    handleKeyDown: (e: React.KeyboardEvent<HTMLButtonElement>) => void;
```
이벤트로 받아올 땐 handle대신 on을 사용해주면 어떨까요!?

`onClick={handleClick}`

요로코롬 쓰일 수 있도록..

JunilHwang commented on src/components/tabMenu/1.displayAttribute/index.tsx:
```diff
@@ -0,0 +1,94 @@
+            />
+          ))}
+        </div>
+        <TabPanel
+          id={data[currentIndex].id}
+          content={data[currentIndex].content}
+          current={true}
+        />
```
아마 Panel이 현재 탭메뉴에 대한 콘텐츠인 것 같은데
콘텐츠를 바깥에서 건내줄 수 있도록 해주면 어떨까요!?

JunilHwang replied:
Compound Component 처럼 이렇게 쓰이는 모습도 상상해보시면 좋을 것 같아요!

```tsx
<Tab.MenuList>
  <Tab.Menu>메뉴1<Tab.Menu>
  <Tab.Menu>메뉴2<Tab.Menu>
  <Tab.Menu>메뉴3<Tab.Menu>
</Tab.MenuList>

<Tab.Content>
  콘텐츠 1
</Tab.Content>
<Tab.Content>
  콘텐츠 2
</Tab.Content>
<Tab.Content>
  콘텐츠 3
</Tab.Content>
```

참고링크: https://www.chakra-ui.com/docs/components/tabs

ericagong replied:
앗! 이 부분은 안그래도 step1이 다 끝나면 compound pattern으로 전체 코드 리팩토링을 하려고 했어요!

JunilHwang commented on src/components/tabMenu/2.selectiveRender/index.tsx:
```diff
@@ -0,0 +1,83 @@
+    handleClick: () => void;
+    handleKeyDown: (e: React.KeyboardEvent<HTMLButtonElement>) => void;
+  }
+>(({ title, current, handleClick, handleKeyDown }, ref) => {
```
앞선 리뷰와 동일합니다!

JunilHwang commented on src/components/textBox/2_clone/index.tsx:
```diff
@@ -0,0 +1,70 @@
+    return () => {
+      if (elem) elem.removeEventListener('input', handleInput);
+    };
+  }, []);
```
```suggestion
  useEffect(() => {
    const elem = textareaRef.current;
    const cloneElem = cloneRef.current;
    if (!elem) {
        return undefined;
    }

    const handleInput = () => {
      if (!cloneElem) return;

      const value = elem.value;
      cloneElem.value = value;

      const totalRows = Math.floor(
        cloneElem.scrollHeight / cloneElem.clientHeight,
      );
      const clampedRows = clamp(totalRows, MIN_LINES, MAX_LINES);

      elem.rows = clampedRows;
    };

    elem.addEventListener('input', handleInput);

    return () => elem.removeEventListener('input', handleInput);
  }, []);
```
이렇게 최상단에서 미리 필터링 해버리면 뒤에서 안 해줘도 된답니다!

JunilHwang commented on src/components/lineClamp/1_canvas/index.tsx:
```diff
@@ -0,0 +1,65 @@
+      const measuredLines = calculateLines(elemRef.current, text);
+      setIsClamped(measuredLines > maxLines);
+    }
+  }, [text, maxLines]);
```
canvas를 통해 계산하는 방식이군요..!

JunilHwang replied:
https://tailwindcss.com/docs/line-clamp

사실 이런거 쓰면 간단하게 해결 되긴해요 ㅋㅋ

```tsx
.line-clamp-3 {
    -webkit-line-clamp: 3;
    -webkit-box-orient: vertical;
    display: -webkit-box;
    overflow: hidden;
}
```

요로코롬...

JunilHwang commented on src/components/tooltip/1_detailsSummary/index.tsx:
```diff
@@ -0,0 +1,86 @@
+        <h3>#1. Click 이벤트 기반 툴팁(details summary 태그 기반 구현)</h3>
+        {data.map((d) => (
+          <Tooltip {...d} key={d.id} />
+        ))}
```
아무데나 클릭해도 뜨고 있는데, 의도하신 부분이겠쬬..!?

JunilHwang commented on src/components/tooltip/2_clickEventHandler/index.tsx:
```diff
@@ -0,0 +1,96 @@
+          aria-expanded={isActive}
+          aria-describedby={isActive ? tooltipId : undefined}
+          tabIndex={0}
+        ></button>
```
여기서는 트리거를 기반으로 동작하도록 만들어줬네요 ㅎㅎ

JunilHwang reviewed (commented):
일단 1차 피드백 남겨놨습니다! 멘토링 하면서 또 이야기나눠봐요~!

ericagong committed:
refactor: 스크롤 관련 UI 요소 1차 피드백 반영

ericagong committed:
feat: Snackbar 구현

  핵심 구현 사항
  - App 수준에서 스낵바 상태 관리 위해: Context API OR React.createPortal 사용
  - 사용자 스낵바 hover 시 스낵바 열림 상태 유지
  - 스낵바 닫힘 버튼 클릭 시 닫힘 처리

  최종 선택안: React.createPortal 기반 구현
  - 로컬에서 상태 제어 + 전역 UI 출력에 최적화된 creatPortal 방식 채택
  - 한계: createPortal($el, $container) -> $container 반드시 마운트 이후 접근 -> useEffect 처리 필요
  - SSR 환경에서는 브라우저 환경 보장 전까지 동작 불가

ericagong committed:
feat: 모달(modal) 구현

ericagong committed:
feat: popover 구현

ericagong committed:
feat: 이미지 슬라이드 구현

ericagong committed:
feat: carousel 구현

ericagong committed:
feat: 드롭다운(selectBox) 구현

ericagong committed:
feat: Drag and Drop 리스트 구현

ericagong committed:
chore: 개발 환경 설정(ESLint + 패키지 버전업)

ericagong committed:
chore: 개발 환경 설정(prettier)

ericagong committed:
chore: 개발 환경 설정(stylelint)

ericagong committed:
chore: 개발 환경 설정(husky + lint-staged)

ericagong committed:
chore: 개발 환경 설정

ericagong committed:
refactor: 아코디언 컴포넌트 확장

ericagong committed:
docs: 아코디언 컴포넌트 리팩토링 방향성 수립

ericagong committed:
refactor: 탭메뉴(Tabs) Compound + Headless UI로 변환 및 로직과 뷰 분리

ericagong committed:
refactor: 단일 열림 정책 아코디언

ericagong committed:
refactor: 아코디언 컴포넌트 구현

- Compound + Headless UI 패턴으로 뷰와 기능 분리
- 핵심 UI 기능 customHook 분리: ctrl+F 브라우저 이벤트 연동
- 추가 리팩토링 방향성: Controlled Accordion으로 확장

ericagong committed:
Revert "refactor: 아코디언 컴포넌트 구현"

This reverts commit 13898e0a028bc1fbaadf55fbedd6f835d9886498.
